src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,231 @@
+      mmon_component &operator = (mmon_component &&) = delete;
+
+      /* function */
+      const char *get_name ();
+      void add_stat (uint64_t size, int subcomp_idx, bool init, bool expand);
+      void sub_stat (uint64_t size, int subcomp_idx, bool init);
+      int get_subcomp_index (const char *subcomp_name);
+      int add_subcomponent (const char *name);
+
+    private:
+      std::string m_comp_name;
+      MMON_STAT m_stat;
+      std::vector<std::unique_ptr<mmon_subcomponent>> m_subcomponent;
+  };
+
+  class mmon_module
+  {
+    public:
+      mmon_module (const char *module_name, const MMON_COMP_INFO *info);
+      mmon_module () {} /* for dummy */
+      mmon_module (const mmon_module &) = delete;
+      mmon_module (mmon_module &&) = delete;
+
+      virtual ~mmon_module () {}
+
+      mmon_module &operator = (const mmon_module &) = delete;
+      mmon_module &operator = (mmon_module &&) = delete;
+
+      /* function */
+      virtual int aggregate_stats (const MMON_MODULE_INFO &info);
+      void add_stat (uint64_t size, int comp_idx, int subcomp_idx, bool init, bool expand);
+      void sub_stat (uint64_t size, int comp_idx, int subcomp_idx, bool init);
+      int add_component (const char *name);
+      inline int MAKE_STAT_IDX_MAP (int comp_idx, int subcomp_idx);
+
+    private:
+      /* const */
+      /* max index of component or subcomponent is 0x0000FFFF
+       * if some stats have max index of component or subcomponent,
+       * we don't have to increase it */
+      static constexpr int MAX_COMP_IDX = 0x0000FFFF;
+
+    private:
+      std::string m_module_name;
+      MMON_STAT m_stat;
+      std::vector<std::unique_ptr<mmon_component>> m_component;
+      std::vector<int> m_comp_idx_map;
+  };
+
+
+  class memory_monitor
+  {
+    public:
+      memory_monitor (const char *server_name)
+	: m_aggregater (this), m_server_name (server_name) {}
+      memory_monitor (const memory_monitor &) = delete;
+      memory_monitor (memory_monitor &&) = delete;
+
+      ~memory_monitor () {}
+
+      memory_monitor &operator = (const memory_monitor &) = delete;
+      memory_monitor &operator = (memory_monitor &&) = delete;
+
+      /* function */
+      int add_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID stat_id, uint64_t size, bool expand);
+      int sub_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID stat_id, uint64_t size);
+      int resize_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID stat_id, uint64_t old_size, uint64_t new_size);
+      int move_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID src, MMON_STAT_ID dest, uint64_t size);
+      int aggregate_module_info (MMON_MODULE_INFO *info, int module_index);
+      int aggregate_tran_info (MMON_TRAN_INFO *info, int tran_count);
+      inline int COMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val);
+      inline int SUBCOMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val);
+
+    private:
+      class aggregater
+      {
+	public:
+	  aggregater (memory_monitor *mmon);
```

Comment:
Nit:

About memory_monitor *mmon

Pointer -> Reference.

Reply:
I previously attempted to address the same comment raised by wrlawodms, but I have confirmed that it is not possible to set the backlink within a member variable of the class in that way.